### PR TITLE
[SMD-491] introduce IngestConfig

### PR DESCRIPTION
**Pull request by cyrusdobbs**
### Change description
- introduces a new class `IngestConfig` that encapsulates the various dependencies that are required to ingest a round of data
- `IngestConfig` injects dependencies depending on the fund and round specific in the request to `/ingest`

- [X] Unit tests and other appropriate tests added or updated
- [ ] README and other documentation has been updated / added (if needed)
- [X] Commit messages are meaningful and follow good commit message guidelines (e.g. "FS-XXXX: Add margin to nav items preventing overlapping of logo")

### How to test
_If manual testing is needed, give suggested testing steps_

### CODE REVIEW

File: core/injection.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+from core.validation_schema import (
+    ROUND_FOUR_TF_SCHEMA,
+    ROUND_ONE_TF_SCHEMA,
+    ROUND_THREE_TF_SCHEMA,
+    ROUND_TWO_TF_SCHEMA,
+)
+
+
+@dataclass
+class IngestConfig:
```
Comment (cyrusdobbs) [inline core/injection.py:27]: Not sure I like this naming.. `IngestDependencies` maybe better?

**Review by Connor-OS — COMMENTED** [review]
I think this makes onboarding a new fund much clearer and the code more self documenting

File: core/controllers/ingest.py [hunk 2]
```diff
@@ -59,67 +59,90 @@ def ingest(body: dict, excel_file: FileStorage) -> Response:
     :raises ValidationError: raised if the data fails validation
     """
     g.excel_file = excel_file.stream
+    fund = body.get("fund", "Towns Fund")  # defaults to Towns Fund
```
[the remark this answers is not in the document]
Comment (cyrusdobbs) [reply core/controllers/ingest.py:62]: That was my thinking. However seeing as though the front-end is already set up to provide this, I actually think I'll remove this default, add the param to the OpenAPI docs, and correct the tests to pass it as well.

File: core/controllers/ingest.py [hunk 3]
```diff
@@ -59,67 +59,90 @@ def ingest(body: dict, excel_file: FileStorage) -> Response:
...
 
+    ingest_config: IngestConfig = INGEST_CONFIGS.get(fund, reporting_round)
+
+    original_workbook = extract_data(excel_file=excel_file)
     try:
-        pre_transformation_validations(original_workbook, reporting_round, auth)
-        workbook = transform_data(original_workbook, reporting_round)
-        validate(workbook, original_workbook, reporting_round)
+        if ptv_schema := ingest_config.pre_transformation_validation_schema:
+            pre_transformation_validations(original_workbook, ptv_schema, auth)
```
[the remark this answers is not in the document]
Comment (cyrusdobbs) [reply core/controllers/ingest.py:72]: Yep we should be able to. Although that may not be true if we plan to ingest collections of submissions like R1 and R2 again in the future - as opposed to individual submission files

File: core/injection.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    def __init__(self, fund_round_to_fund_configs: dict[tuple[str, int], IngestConfig]):
+        self._fund_configs = fund_round_to_fund_configs
+
+    def get(self, fund: str, reporting_round: int) -> IngestConfig:
+        """Retrieve the IngestConfig mapped to a fund and reporting round.
+
+        :param fund: fund name
+        :param reporting_round: reporting round
+        :raises ValueError: if the given fund and reporting_round does not match an IngestConfig
+        :return:
```
[the remark this answers is not in the document]
Comment (cyrusdobbs) [reply core/injection.py:63]: nice spot

File: core/injection.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
+"""
+Module responsible for handling dependency injection to the ingest endpoint.
```
[the remark this answers is not in the document]
Comment (cyrusdobbs) [reply core/injection.py:2]: yeah good points. Going to move to `ingest.py` for now

File: core/controllers/ingest.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cyrusdobbs) [reply core/controllers/ingest.py]: I've removed this default and added it to the OpenAPI spec now 👍

File: core/injection.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cyrusdobbs) [reply core/injection.py]: Didn't like the mass of imports that this then causes `ingest.py` to load.. moved it to `ingest_dependencies.py` within controllers. Feels a bit odd but I think it's best to just keep it close to `ingest.py` for clarity

**Review by Connor-OS — APPROVED** [review]
LGTM, really like these changes 👍

**Pull request merged**